Compiler team 2022 midyear report (#1017)
rust-lang/blog.rust-lang.org: posts/inside-rust/2022-07-29-compiler-team-2022-midyear-report.md

goals* or *no plan* to achieve those goals. Ten of those fourteen say they now
have a better understanding of the problem, and five of those fourteen say they
now have a plan.

  * Here is a question I want to put forward to our communtiy: Do we, as a
    project, need to put more energy into upfront planning? Or, given our
    volunteer nature, will that just distract people from "the fun stuff" and
    potentially burn out contributors? In other words: is this 5/14 "ambitions
    that spent time making a concrete plan", is that a value we
    should work on growing? Or should we just monitor it, and not actively try to
    change it?

  * Note: It is possible that my survey was itself somewhat flawed, so we need
    to be careful in how we interpret this data. My intention was to try to
    capture how well structured a given ambition's plan was by asking if it had
    established milestones and/or a schedule. However, the survey results
    include cases where the respondant did *not* indicate that they had made a
    plan, but they *did* say that this year they did establish milestones.
    Should I have interpreted that as implicitly saying that there *was* a plan

>     established? Maybe we need to restructure future surveys to make these

Reviewer: It could be interpreted as retroactively saying that something was a big enough improvement to call it hitting a milestone, without having a concrete plan before those milestones were achieved. And that there is no plan on what milestones to hit next.

Author: exactly, that is an excellent example of the kind of way that one might have had very good reason to click the checkbox saying that milestones were hit while still leaving the other checkboxes unset.

I'll incorporate your example into the text

Author: Added in https://github.com/rust-lang/blog.rust-lang.org/pull/1017/commits/0f0bf0239804a895fde8da65dbf9d5315d3d8e75

Mentoring instructions, or even steps with minimal details, might help non-experts pick up work in the area.

#### Cranelift

<!-- https://hackmd.io/@rust-compiler-team/BJLre2Xu9 -->

**Progress:** The Rust project has made progress on this ambition since the start of the year.

**Goals:** We had no goals for this ambition planned for this year, but we made ad-hoc progress on the problem itself.
With regards to the overall problem: In terms of getting cg_clif distributed with rustup, @bjorn3 thinks 6 months is feasible. In terms of implementing all missing rust features they think it will take longer.

**How it started:** at start of 2022, we had a plan, some of which had been completed, but more work remained to be done.

**How it's going:** we have a high-level end-to-end plan to solve the problem, and we have completed some of our milestones.

**Regarding new contributors:** There have been a couple minor PR’s. There has also been someone potentially interested in contributing more, but given that most of the low hanging fruit has already been done by @bjorn3, they were not able to help the newcomer find an area to contribute to.

> **Regarding prioritization and focus:** @bjorn3 has kept their focus on working on `cg_clif` and rustc PR’s that are part of the time related to `cg_clif`.

Reviewer: This sentence is gramatically incorrect.

Author: You're right.

```suggestion
**Regarding prioritization and focus:** @bjorn3 has kept their focus on working on `cg_clif` and rustc PR’s that are related to `cg_clif`.
```

Is this better?

Reviewer: Yes

</table>

Here are some trends I noted from looking at the table:

The overwhelming majority of our ambitions, 19 out of 22, reported *some*
amount of progress. That's great, especially given that several of those
ambitions were mere aspirations that explicitly had no resources attached to
them at the start of the year.

Less than a quarter of the ambitions, 6 out of 22, said that their overall
problem would be resolved, or the bulk of the problem addressed, within the next
six months. Ten others said they predicted the important components of their
solution would be available within the next two years, so that's a total of 16
out of 22 that expect some kind of solution within two years.

  * Note: One respondent noted that the two year threshold embedded in the
    survey was an "odd timeframe." I do not disagree with that remark. The
    reality is that I was trying to find a happy medium between two extremes:
    short timeframes don't allow enough time for significant work to be
    accomplished, while excessively long timeframes (e.g. greater than five
    years) are very difficult to predict realistically. But the two year
    time frame was drawn somewhat out of a hat.

> Of the nineteen ambitions that reported some progress so far this year, fourteen
> of them (14/19) also said that at the start of the year, they had either *no
> specific goals*, or *no plan* to achieve their goals. Ten of those fourteen say they now
> have a better understanding of the problem, and five of those fourteen say they
> now have a plan.

Author: ```suggestion
Of the nine concrete initiatives (i.e. ambitions with development resources attached) that reported some progress so far this year, four said they had no specific goal at the start of the year, and two said they had no plan to achieve their goal.
Similarly, of the ten aspirations (i.e. ambitions without development resources attached) that reported progress, six had no specific goals, and two said they had no plan to achieve their goals.

&nbsp; | No Specific Goals | No Plan | Other
--------|-------------------|--------|--------
Concrete Initiative | 4 | 2 | 3
Aspiration | 6 | 2 | 2
 
 Four of those six concrete initiatives that lacked either specific goals or a plan say they now
have a better understanding of the problem; two of them say they
now have a plan.

Five of the eight aspirations that lacked either specific goals or a plan say they now better understand the problem, and two now have a plan.

I spell this out in this level of detail because I was *expecting* to see some strong correlation between having development resources and having a plan, at least compared to what the aspirations reported. I think the numbers above show that expectation was incorrect. I'm not yet sure what lesson to draw from this, apart from being happy that so many ambitions report that they have a better understanding of their respective problem.
```

>   * Here is a question I want to put forward to our community: Do we, as a

Author: ```suggestion
  * Here is a question I want to discuss with the compiler contributors in some future [compiler steering meeting](https://rust-lang.github.io/compiler-team/about/steering-meeting/): Do we, as a
```

**Details:**
We've made concrete improvements/fixes to debuginfo generation.

Specifically:
- Change char type in debuginfo to DW_ATE_UTF [#89887](https://github.com/rust-lang/rust/pull/89887)
- Fix debuginfo for pointers/references to unsized types [#93006](https://github.com/rust-lang/rust/pull/93006)
- debuginfo: Support fat pointers to unsized tuples. [#94050](https://github.com/rust-lang/rust/pull/94050)
- debuginfo: Fix bug in type name generation for dyn types with associated types but no other generic arguments. [#94810](https://github.com/rust-lang/rust/pull/94810)
- async: Give predictable name to binding generated from .await expressions. [#95011](https://github.com/rust-lang/rust/pull/95011)
- debuginfo: Fix debuginfo for Box&lt;T&gt; where T is unsized. [#95270](https://github.com/rust-lang/rust/pull/95270)
- debuginfo: Emit ZST struct debuginfo for unit type when CPP-like debuginfo is enabled [#96316](https://github.com/rust-lang/rust/pull/96316)

Surprises:
Debuginfo just doesn't have enough test coverage, but that isn't particularly surprising.

**Regarding prioritization and focus:** debugging in general is a top priority for mw & wesleywiser’s

> team. As pure conjecture, if they were not working on debugging, then they might

Author: Fixed in 9e9bc5a95a8f20c3e641cccf1140bed353dc7965

be working on compiler performance or a long tail of random features that are
either not implemented or unstable.

### supporting split debuginfo

<!-- https://hackmd.io/@rust-compiler-team/ByXfjiXu5 -->

**Progress:** The Rust project has not made any progress on this ambition since the start of the year (pnkfelix: but the free form answer somewhat contradicts this)

**Goals:** As of today, we think this year’s planned goals for this ambition will be achieved in the next six months, and we think the most important parts of a solution will be available in the next six months.

**Where we started:** At the start of 2022, we had completed most of our plan; our main focus was polish and getting the work into the hands of Rust users.

**How it's going:** As of today, we think we have implemented a solution to the problem, but most of our work has not yet landed in hands of Rust users

> **Details:** There hasn't been a lot of progress on split debuginfo in 2022 because most of the big pending work that was in-progress landed in late 2021 (namely integration of [thorin](https://github.com/rust-lang/thorin) into rustc to support cross-crate Split DWARF). Since then, Split DWARF has been basically complete and just simmering on nightly before stabilization is proposed. Split debuginfo on non-DWARF platforms is already complete and stable.

Author: Fixed in b4fe5ecbcd098818c273bc66f2fa652e1ec18ec2

[PR #92268](https://github.com/rust-lang/rust/pull/92268) is now undergoing the final polish required for it to be merged, and [near-future units of follow-up work](https://rust-lang.zulipchat.com/#narrow/stream/216762-project-safe-transmute/topic/Implementation/near/290258987) have been identified.

**Regarding new contributors:** An influx of interest and help from [@m1el](https://github.com/m1el) ([Igor null](https://rust-lang.zulipchat.com/#narrow/stream/216762-project-safe-transmute/sender/498326-user498326) on Zulip) jolted Project Safe Transmute out of its doldrums. Additionally, [@joshlf](https://github.com/joshlf), an early collaborator on Project Safe Transmute, anticipates he will soon be able to rejoin the implementation effort.

**Regarding prioritization and focus:** Personal and professional obligations sapped the capacity of collaborators to contribute. These obligations have been resolved, and progress is being made once again.

> @jswrenn's work on as-casting and enum reform has been slowed by my focus on safe transmute. @jswrenn is increasingly able to devote attention to enum reform, but most of my attention remains on project safe transmute. They've allocated most of my attention on safe transmute, because they believe it is critically important to writing safe, performant code.

Author: Whoops, I'll fix that.

> @jswrenn's work on as-casting and enum reform has been slowed by my focus on safe transmute. @jswrenn is increasingly able to devote attention to enum reform, but most of my attention remains on project safe transmute. They've allocated most of my attention on safe transmute, because they believe it is critically important to writing safe, performant code.

Author: fixed in da4164dc14bdcb6aa9f612c0d521e2e9f72f0a0c

we have Rust contributors who have agreed to help with the units of work that we have identified

Details:

We have refactored the global HIR visitor to reduce the amount of accesses to HIR. Using this, we managed to reduce incremental reuse and avoid some recomputations.

We have laid down the first steps for implementing “salsa jars” in the rustc query system. Those first steps are being implemented by a contributor, but the future plans are not fully designed yet.

We have established the constraints for extending the query system with “salsa entities”. The design is much more subtle than anticipated, and will take longer than planned.

Integration of lowering into the query system is still in review. This blocks progress on making name resolution incremental.

Regarding new contributors:
The progress has been made in part thanks to Miguel Guarniz (kckeiks).

Regarding prioritization and focus:
One of the owners, cjgillot started a large refactor of lifetime resolution.
This refactor allowed for faster progress in fixing a few old bugs.

### Inter-crate Sharing Aspirations

<!-- https://hackmd.io/zad1BjIoR8S4IfE49z-xYg -->
<!-- (no survey response as of 2022-07-19) -->

> No survey response was provided; we assume no progress has been made. This aspiration may have been abandoned.

Reviewer 2: I think we should follow up and confirm this so we can make a definitive statement.

Author: Okay I'll reach out. (I think I might have already, but I'll double-check in any case.)

Author: Fixed in 18fd548b7da90bfb8cf9c6c84467c96d97c1d370

there is specific feature and integration work that can be tackled in bounded amount of times, but the improvement of diagnostics is a fractal one, and every time we raise the bar, the expectation of our users increases and our increased experience makes us attempt bolder things.

**How it started:** at start of 2022, we knew some problems existed, but we did not yet have a specific goal in mind for solving all such problems.
for other problems, we had a plan for how to achieve a specific goal, but we did not have intermediate progress points (aka “milestones”) established.

**How it's going:** we understand the problem better than we did at the start of the year
we have a high-level end-to-end plan to solve the problem
we have milestones that represent individual units of work for the near term that each make progress towards a solution
we have Rust contributors who have agreed to help with the units of work that we have identified.
Further more, some of our work has reached Rust programmers; in some cases, we do not know if it has improved Rust for them,
and in other cases, what we learn of their usage is informing our plans going forward

> **Details:** wg-diagnostics isn't as structured as it could be, which translates to work being taken on piecemeal by individuals as at their own pace and leisure. This affects the size of projects that we can tackle, *but* still relatively big projects, like the [translation machinery](https://hackmd.io/@davidtwco/rkXSbLg95) were started and have enough momentum to .

Author: Fixed in 734b8c60497b4d90f2a0e911fc669267813d94af

---
layout: post
title: Rust Compiler Midyear Report for 2022
author: Felix Klock, Wesley Wiser
description: "The compiler team's midyear report on its ambitions for 2022."
team: The Compiler Team <https://www.rust-lang.org/governance/teams/compiler>
---

# Rust Compiler Midyear Report for 2022

Back in February, the compiler team [posted a collection][ambitions-post] of
concrete initiatives and hopeful aspirations for this year. This post is a
midyear report summarizing the progress so far on all of those items.

[ambitions-post]: /inside-rust/2022/02/22/compiler-team-ambitions-2022.html

(The original plan was to have this report out by the end of June, but there

> were extenuating circumstances that delayed the post until now.)

Reviewer 2: I would consider dropping this sentence. It doesn't really add to the content of the post and we're only about 1-2 weeks off from the halfway point of 2/22 - 12/31 anyway.

Author: ```suggestion
```

ambitions were mere aspirations that explicitly had no resources attached to
them at the start of the year.

Less than a quarter of the ambitions, 6 out of 22, said that their overall
problem would be resolved, or the bulk of the problem addressed, within the next
six months. Ten others said they predicted the important components of their
solution would be available within the next two years, so that's a total of 16
out of 22 that expect some kind of solution within two years.

  * Note: One respondent noted that the two year threshold embedded in the
    survey was an "odd timeframe." I do not disagree with that remark. The
    reality is that I was trying to find a happy medium between two extremes:
    short timeframes don't allow enough time for significant work to be
    accomplished, while excessively long timeframes (e.g. greater than five
    years) are very difficult to predict realistically. But the two year
    time frame was drawn somewhat out of a hat.

Of the nineteen ambitions that reported some progress so far this year, fourteen
of them (14/19) also said that at the start of the year, they had either *no

> goals* or *no plan* to achieve those goals. Ten of those fourteen say they now

Reviewer 2: I think this framing is not quite accurate. From what I can see, every initiative/aspiration has some kind of goal (stabilize feature x, improve problem y, etc), it was having a concrete plan to achieve the goal that was lacking. Many of the ambitions are open ended as well and another interpretation of the data for those groups is that "there is no plan that will achieve the goal because it is never ending" (ie, we can always make diagnostics/debuginfo/compiler perf/etc better).

Author: I had a bunch more private conversation with Wesley about this and I think we're okay with the text after the subsequent revisions we made on 2022-08-05.

> of them (14/19) also said that at the start of the year, they had either *no
> goals* or *no plan* to achieve those goals. Ten of those fourteen say they now

Author: ```suggestion
of them (14/19) also said that at the start of the year, they had either *no
specific goals*, or *no plan* to achieve their goals. Ten of those fourteen say they now
```

have a better understanding of the problem, and five of those fourteen say they
now have a plan.

>   * Here is a question I want to put forward to our community: Do we, as a

Reviewer 2: How should we collect responses? Also, which community are we asking? The community of compiler contributors, project contributors, Rust users, etc.

Author: The "how should we collect responses" is a good point.

I guess I'm asking the community of compiler contributors.

Reviewer 2: I think that's a good topic for a steering meeting then. As it is, it's likely this will get read by a much larger audience than just the compiler contributors and the post sounds as if we want feedback from the whole world 😄

Author: Hmm. Okay. I still want to make it clear in the text that this is a question that I want to be asking, but I will revise the text to reframe the question as one that I will be trying to resolve in collaboration with the compiler-contributors.

    capture how well structured a given ambition's plan was by asking if it had
    established milestones and/or a schedule. However, the survey results
    include cases where the respondent did *not* indicate that they had made a
    plan, but they *did* say that this year they did establish milestones.
    Should I have interpreted that as implicitly saying that there *was* a plan
    established? Or is the respondent saying that something significant was
    accomplished despite the absence of a plan with concrete milestones?
    Maybe we need to restructure future surveys to make these
    "implied bounds" clearer (ha ha, that's a little Rust type system joke).

One more thing I want to point out: five of the ambitions checked the box in the
survey that said "some of our work has reached Rust programmers, but we do not
know if it has improved Rust for them." (This is listed as "need-user-feedback"
in the table above.) We should figure out ways to provide such feedback. I know
that in one case, the lead for diagnostic improvements scours twitter to see if
people are complaining about Rust's error messages (or celebrating them!).

> That's one approach, but I think we have to figure out something better. Perhaps
> we could provide an IDE plugin that provide telemetry to the project about the

Reviewer 2: Given how controversial collecting telemetry is, we should at least throw "opt-in" here somewhere.

```suggestion
That's one approach, but I think we have to figure out something better. Perhaps
we could provide an opt-in plugin that provide telemetry to the project about the
```

**How it started:** At start of 2022, we knew a problem existed, but we did not yet have a specific goal in mind for solving the problem.

**How it's going:** We would like help deciding what to do next.
We understand the problem better than we did at the start of the year,  we have Rust contributors who have agreed to help with the units of work that we have identified,
and we have completed some of our milestones. Furthermore, (some of) our work has reached Rust programmers. In some cases, we do not know if it has improved Rust for them; in others, what we learn of their usage is informing our plans going forward.

**Details:** wg-debugging has been spun up. We now are conducting regular status/design and triage meetings. We are working through the backlog of A-debuginfo issues. There is active participation from a number of contributors.

### improving debuginfo quality

<!-- https://hackmd.io/@rust-compiler-team/HJQ25sXOq -->

**Progress:** The Rust project has made progress on this ambition since the start of the year

**Goals:** We had no goals for this ambition planned for this year, but we made ad-hoc progress on the problem itself.

> There’s nearly an unbounded amount of effort that could be spent improving debuginfo quality but wesleywiser thinks we are making significant improvement both over the last 6 months and in the final 6 months of this year as well.

Reviewer 2: Nit: name style consistency

```suggestion
There’s nearly an unbounded amount of effort that could be spent improving debuginfo quality but @wesleywiser thinks we are making significant improvement both over the last 6 months and in the final 6 months of this year as well.
```

**How it started:** At start of 2022, we knew a problem existed, but we did not yet have a specific goal in mind for solving the problem.

**How it's going:** We would like help deciding what to do next.
We understand the problem better than we did at the start of the year,  we have Rust contributors who have agreed to help with the units of work that we have identified,
Furthermore, (some of) our work has reached Rust programmers. In some cases, we do not know if it has improved Rust for them; in others, what we learn of their usage is informing our plans going forward.

> Much of the work wesleywiser is aware of has landed in 1.60 or 1.61 but there are a few small pieces landing in 1.62 (current beta).

Reviewer 2: Nit

```suggestion
Much of the work @wesleywiser is aware of has landed in 1.60 or 1.61 but there are a few small pieces landing in 1.62 (current beta).
```

**Details:**
We've made concrete improvements/fixes to debuginfo generation.

Specifically:
- Change char type in debuginfo to DW_ATE_UTF [#89887](https://github.com/rust-lang/rust/pull/89887)
- Fix debuginfo for pointers/references to unsized types [#93006](https://github.com/rust-lang/rust/pull/93006)
- debuginfo: Support fat pointers to unsized tuples. [#94050](https://github.com/rust-lang/rust/pull/94050)
- debuginfo: Fix bug in type name generation for dyn types with associated types but no other generic arguments. [#94810](https://github.com/rust-lang/rust/pull/94810)
- async: Give predictable name to binding generated from .await expressions. [#95011](https://github.com/rust-lang/rust/pull/95011)
- debuginfo: Fix debuginfo for Box&lt;T&gt; where T is unsized. [#95270](https://github.com/rust-lang/rust/pull/95270)
- debuginfo: Emit ZST struct debuginfo for unit type when CPP-like debuginfo is enabled [#96316](https://github.com/rust-lang/rust/pull/96316)

Surprises:
Debuginfo just doesn't have enough test coverage, but that isn't particularly surprising.

> **Regarding prioritization and focus:** debugging in general is a top priority for mw & wesleywiser’s

Reviewer 2: Nit

```suggestion
**Regarding prioritization and focus:** debugging in general is a top priority for @mw & @wesleywiser’s
```

team.

### supporting split debuginfo

<!-- https://hackmd.io/@rust-compiler-team/ByXfjiXu5 -->

> **Progress:** The Rust project has not made any progress on this ambition since the start of the year (pnkfelix: but the free form answer somewhat contradicts this)

Reviewer 2: Nit

```suggestion
**Progress:** The Rust project has not made any progress on this ambition since the start of the year (@pnkfelix: but the free form answer somewhat contradicts this)
```

**How it's going:** As of today, we think we have implemented a solution to the problem, but most of our work has not yet landed in hands of Rust users

**Details:** There hasn't been a lot of progress on split debuginfo in 2022 because most of the big pending work that was in-progress landed in late 2021 (namely integration of [thorin](https://github.com/rust-lang/thorin) into rustc to support cross-crate Split DWARF). Since then, Split DWARF has been basically complete and just simmering on nightly; an FCP for stabilization on Linux has completed. Split debuginfo on non-DWARF platforms is already complete and stable.

Progress this year has primarily been some benchmarking of Split DWARF and some work to make split debuginfo an option when bootstrapping rustc (i.e. in `config.toml`), but not a lot else.

Future work is basically just stabilization of `-Csplit-debuginfo` on Linux (Split DWARF); and of the currently-default options for the other platforms (for example, `-Csplit-debuginfo=packed` on Windows requires `-Zunstable-options` despite being effectively the default if you don't specify any flags).

> The owner of this work, davidtwco, intends to stick with the theme of debugging and contribute to the wg-debugging working group, but has also shifted attention to diagnostic translation they see that as an interesting area where they can have impact (and because the remaining implementation tasks for split debuginfo were completed as noted above).

Reviewer 2: Nit

```suggestion
The owner of this work, @davidtwco, intends to stick with the theme of debugging and contribute to the wg-debugging working group, but has also shifted attention to diagnostic translation they see that as an interesting area where they can have impact (and because the remaining implementation tasks for split debuginfo were completed as noted above).
```

### better integration with trace-based debuggers

<!-- https://hackmd.io/@rust-compiler-team/BkJSsi7d5 -->

**Progress:** The Rust project has not made any progress on this ambition since the start of the year.

**Goals:** We do not think this year’s planned goals for this ambition will be achieved in the next six months;
we think a solution will be available in the next two years, but not sooner than six months.

**How it started:** At start of 2022, we had a solution or goal in mind, but did not have a plan for how to achieve it.

**How it's going:** We would like help deciding what to do next.

> **Details:** pnkfelix spent a significant portion of 2021 learning about `rr` and

Reviewer 2: Nit

```suggestion
**Details:** @pnkfelix spent a significant portion of 2021 learning about `rr` and
```

`pernos.co`. They had hoped to spend some of 2022 trying to improve the

> experience when using those tools with Rust, but so far pnkfelix has failed to

Reviewer 2: Nit

```suggestion
experience when using those tools with Rust, but so far @pnkfelix has failed to
```

**Regarding new contributors:** There has not been much community involvement in the implementation, but many people have shown up to express their support (or dissent) in the stabilization PR.

**Regarding prioritization and focus:** Since the opening of the stabilization PR and following pushback, progress has been slow. That has, in part, been to try to incorporate work from other projects (NLL, a-mir-formality) into the “stabilization package” - either through direct improvements (from NLL) or a more clear future (through modeling of GATs in a-mir-formality). However, there are other bits of work (writing docs, triaging new issues) that could be done in parallel that have been somewhat partially neglected.

> For Jack Huey, switching to getting NLL stabilized was a nice change of pace. In a sense, it was “low-hanging fruit” and was a helpful mental break from pushing so hard on GATs for the past year.

Reviewer 2: Nit: I think for consistency we should use a username here

```suggestion
For @jackh726, switching to getting NLL stabilized was a nice change of pace. In a sense, it was “low-hanging fruit” and was a helpful mental break from pushing so hard on GATs for the past year.
```

**How it's going:** we understand the problem better than we did at the start of the year,
we have a high-level end-to-end plan to solve the problem,
we have milestones that represent individual units of work for the near term that each make progress towards a solution
we have Rust contributors who have agreed to help with the units of work that we have identified.
Further more, we have completed some of our milestones; we think we have implemented a solution to the problem, but most of our work has not yet landed in hands of Rust users

**Details:** At the start of the year, we opened [PR #92268](https://github.com/rust-lang/rust/pull/92268), *Initial Implementation of Transmutability Trait*, which aimed to provide the basic functionality of a trait implemented for any two types transmutable into each other (as defined by [MCP #411](https://github.com/rust-lang/compiler-team/issues/411)). This PR required additional testing and polish before it would be ready to merge, but progress unfortunately stalled in the spring.

> With the mentoring provided by [@oli-obk](https://github.com/oli-obk) and an influx of interest and help from [@m1el](https://github.com/m1el), progress resumed this summer; notably:

Reviewer 2: Nit: I think we should either link everyone's usernames or no ones. Doing the easy thing for now but I can generate links if you think we should go that way. 

```suggestion
With the mentoring provided by @oli-obk and an influx of interest and help from @m1el, progress resumed this summer; notably:
```

- A significant effort in testing revealed flaws in the initial implementation approach. Fortunately, we quickly [discovered](https://rust-lang.zulipchat.com/#narrow/stream/216762-project-safe-transmute/topic/Implementation/near/288584316) and implemented an alternative (and arguably simpler) implementation strategy!
- The `rustc_transmute` crate now only *optionally* depends on other `rustc_*` dependencies, allowing contributors to edit, build, and test the core implementation using the familiar `cargo` commands, rather than building the entire compiler.

[PR #92268](https://github.com/rust-lang/rust/pull/92268) is now undergoing the final polish required for it to be merged, and [near-future units of follow-up work](https://rust-lang.zulipchat.com/#narrow/stream/216762-project-safe-transmute/topic/Implementation/near/290258987) have been identified.

> **Regarding new contributors:** An influx of interest and help from [@m1el](https://github.com/m1el) ([Igor null](https://rust-lang.zulipchat.com/#narrow/stream/216762-project-safe-transmute/sender/498326-user498326) on Zulip) jolted Project Safe Transmute out of its doldrums. Additionally, [@joshlf](https://github.com/joshlf), an early collaborator on Project Safe Transmute, anticipates he will soon be able to rejoin the implementation effort.

Reviewer 2: Nit: My understanding is that m1el/Igor Null are the same person so I'm removing that part of the sentance.

```suggestion
**Regarding new contributors:** An influx of interest and help from @m1el jolted Project Safe Transmute out of its doldrums. Additionally, @joshlf, an early collaborator on Project Safe Transmute, anticipates he will soon be able to rejoin the implementation effort.
```

  we think we have implemented a solution to the problem, but most of our work has not yet landed in hands of Rust users.

**Details:** We had three specific aspirations identified in the compiler team blog post:

1. Better integration between Cargo and debuggers, specifically in regards to starting and configuring the debugger. 
   - We have not made progress on this aspiration.
2. Improve expression evaluation support in various debuggers.
   - wg-debugging has had discussions during our status & design meetings on how this could be achieved, but we have not made concrete progress on this aspiration.
3. Revise our debugger extension architecture for rendering Rust data structures.
   - [RFC 3191](https://github.com/rust-lang/rfcs/pull/3191) has been approved and implemented on nightly toolchains but not yet stabilized.

Regarding new contributors:
RFC 3191 was written and implemented by a new contributor (@ridwanabdillahi).

Regarding prioritization and focus:
There isn't much focus happening here currently.

### Faster Builds Aspirations (👩‍💻, 🛠️)

### Incremental Compilation Aspirations

<!-- https://hackmd.io/@rust-compiler-team/B1D7CiX_c -->

> Progress:

Reviewer 2: ```suggestion
**Progress:**
```

The Rust project has made progress on this ambition since the start of the year

Goals:
we had no goals for this ambition planned for this year, but we made ad-hoc progress on the problem itself.
we think a solution will be available in the next two years, but not sooner than six months.

> How it started:

Reviewer 2: ```suggestion
**How it started**:
```

at start of 2022, we knew a problem existed, but we did not yet have a specific goal in mind for solving the problem.

> How it's going:

Reviewer 2: ```suggestion
**How it's going**:
```

we would like help deciding what to do next
we have Rust contributors who have agreed to help with the units of work that we have identified

> Details:

Reviewer 2: ```suggestion
**Details:**
```

We have refactored the global HIR visitor to reduce the amount of accesses to HIR. Using this, we managed to reduce incremental reuse and avoid some recomputations.

We have laid down the first steps for implementing “salsa jars” in the rustc query system. Those first steps are being implemented by a contributor, but the future plans are not fully designed yet.

We have established the constraints for extending the query system with “salsa entities”. The design is much more subtle than anticipated, and will take longer than planned.

Integration of lowering into the query system is still in review. This blocks progress on making name resolution incremental.

Regarding new contributors:

> The progress has been made in part thanks to Miguel Guarniz (kckeiks).

Reviewer 2: Nit

```suggestion
The progress has been made in part thanks to @kckeiks.
```

Regarding prioritization and focus:

> One of the owners, cjgillot started a large refactor of lifetime resolution.

Reviewer 2: Nit

```suggestion
One of the owners, @cjgillot started a large refactor of lifetime resolution.
```

We do not think a complete solution will be achieved in the next two years, but we do think it is a solvable problem
rylev thinks that the real question here is when this becomes something that’s funded and someone wants to focus their work on; that might be never: we’re in an OK state today, but I do think we can do much better (and would likely get some mileage out of it). We’re probably already 75-80% of a ‘perfect’ solution though, I suspect.

**How it started:** at start of 2022, we knew a problem existed, but we did not yet have a specific goal in mind for solving the problem.

**How it's going:** we would like help deciding what to do next
@rylev also wants to emphasize that the largest problem in this domain is room and bandwidth for experimentation.

**Details:** We’re reporting on a few additional metrics in PRs now (cycles, RSS). It’s not really a surprise, but the significant challenge we’ve definitely run into is our audience has such a diverse set of needs that any single representation or comment is likely to be too information dense to be useful; we’re still figuring out how to make the most of the data we have.

> **Regarding new contributors:** We’ve had a few folks return to contributing this year (nnethercote, lqd), and @rylev believes one new person as well (Kobzol). There have been a few others with good contributions but not long-duration tenure.

Reviewer 2: Nit:

```suggestion
**Regarding new contributors:** We’ve had a few folks return to contributing this year (@nnethercote, @lqd), and @rylev believes one new person as well (@Kobzol). There have been a few others with good contributions but not long-duration tenure.
```

> **Regarding prioritization and focus:** Other infrastructure projects have drawn much of Mark’s attention (crater, triagebot) in the last few months.

Reviewer 2: Nit

```suggestion
**Regarding prioritization and focus:** Other infrastructure projects have drawn much of @Mark-Simulacrum's attention (crater, triagebot) in the last few months.
```

there is specific feature and integration work that can be tackled in bounded amount of times, but the improvement of diagnostics is a fractal one, and every time we raise the bar, the expectation of our users increases and our increased experience makes us attempt bolder things.

**How it started:** at start of 2022, we knew some problems existed, but we did not yet have a specific goal in mind for solving all such problems.
for other problems, we had a plan for how to achieve a specific goal, but we did not have intermediate progress points (aka “milestones”) established.

**How it's going:** we understand the problem better than we did at the start of the year
we have a high-level end-to-end plan to solve the problem
we have milestones that represent individual units of work for the near term that each make progress towards a solution
we have Rust contributors who have agreed to help with the units of work that we have identified.
Further more, some of our work has reached Rust programmers; in some cases, we do not know if it has improved Rust for them,
and in other cases, what we learn of their usage is informing our plans going forward

> **Details:** wg-diagnostics isn't as structured as it could be, which translates to work being taken on piecemeal by individuals as at their own pace and leisure. This affects the size of projects that we can tackle, *but* still relatively big projects, like the [translation machinery](https://hackmd.io/@davidtwco/rkXSbLg95) were started and have enough momentum to make progress going forward.

Reviewer 3: Maybe there's a missing word here, or the "as" is unintended ?

```suggestion
**Details:** wg-diagnostics isn't as structured as it could be, which translates to work being taken on piecemeal by individuals at their own pace and leisure. This affects the size of projects that we can tackle, *but* still relatively big projects, like the [translation machinery](https://hackmd.io/@davidtwco/rkXSbLg95) were started and have enough momentum to make progress going forward.
```

**Regarding new contributors:** @estebank hasn't quantified it, but we've had quite a few new contributors recently. Mentoring capacity isn't depleted, *but* if we wanted to increase the number of people contributing we will need more people. The mentoring needs have been highly variable, depending on the individual (and the complexity of the task they take on).

> **Regarding prioritization and focus:** @estebank has seen people come contribute a handful of PRs and disapear, but has not digged deeper into their reasons.

Reviewer 3: ```suggestion
**Regarding prioritization and focus:** @estebank has seen people come contribute a handful of PRs and disappear, but has not dug deeper into their reasons.
```

The most common thing is people picking up a project that’s too advanced for them, which demotivates them. We try to both steer them away beforehands and to closely mentor them as they work on things. A cleaner codebase with more machinery for non-standard things (like typechecking opportunistically in the parser, to give an example of something impossible to do today) would allow some of the things people have tried to be done by almost anyone.

> @estebank believes that diagnostics are super important and everyone's concern. I think that efforts like librarification could unblock very powerful heuristics to massively improve here, but such a project *shouldn't* be started *only* for diagnostics improvements (as there's plenty of things to do already).

Author: ```suggestion
@estebank believes that diagnostics are super important and everyone's concern. They think that efforts like librarification could unblock very powerful heuristics to massively improve here, but such a project *shouldn't* be started *only* for diagnostics improvements (as there's plenty of things to do already).
```

on, and everything that we hoped other people might want to help with.

In hindsight, it is a huge list. Furthermore, I did not provide much guidance as
to how I would want these items prioritized. (I had my own pet projects, such as
MCVE reduction, but if anything I *deprioritized* those.)

My reasoning at that time was that by going broad with the presentation, we
would be more likely to attract newcomers who would say "oh, *that* thing sounds
like something I could help with"; especially by given examples of projects that
don't actually require deep knowledge of rustc's implementation. But I do not
think that experiment worked out overall.

There are some things, like Async Traits and Generic Associated Types, that are
hugely important, but nonetheless the implementation effort is falling on the
shoulders of relatively few people. I think in the future, I would like to try
to get the team and contributors to be a little more focused. E.g. maybe try to
convince people to put aside work on ambition A, in order to get more people
working together on ambition B, and then later there will (hopefully) be more
capacity to address A after B has been addressed. But, since this is open

> source, its very much a matter of how much excitement each ambition can build up

Reviewer 3: ```suggestion
source, it's very much a matter of how much excitement each ambition can build up
```

Aspiration | 6 | 2 | 2
 
 Four of those six concrete initiatives that lacked either specific goals or a plan say they now
have a better understanding of the problem; two of them say they
now have a plan.

Five of the eight aspirations that lacked either specific goals or a plan say they now better understand the problem, and two now have a plan.

I spell this out in this level of detail because I was *expecting* to see some strong correlation between having development resources and having a plan, at least compared to what the aspirations reported. I think the numbers above show that expectation was incorrect. I'm not yet sure what lesson to draw from this, apart from being happy that so many ambitions report that they have a better understanding of their respective problem.

  * Here is a question I want to discuss with the [compiler team and  contributors](https://www.rust-lang.org/governance/teams/compiler) in some future [compiler steering meeting](https://rust-lang.github.io/compiler-team/about/steering-meeting/): Do we, as a
    project, need to put more energy into upfront planning? Or, given our
    volunteer nature, will that just distract people from "the fun stuff" and

>     potentially burn out contributors? In other words: is this 5/14 "ambitions

Author: ```suggestion
    potentially burn out contributors? In other words: is this 4/14 "ambitions
```

I'm not 100% sure where the 5/14 value came from, but the current text and data presentation pushes hard for 2+2=4.

**Regarding new contributors:** There have been a couple minor PR’s. There has also been someone potentially interested in contributing more, but given that most of the low hanging fruit has already been done by @bjorn3, they were not able to help the newcomer find an area to contribute to.

**Regarding prioritization and focus:** @bjorn3 has kept their focus on `cg_clif` and rustc PR’s that are related to `cg_clif`.

> If @bjorn3 hadn’t started working on `cg_clif` they probably wouldn’t have gained as much knowledge of the inner workings of rustc as they have know, nor contributed much to rustc. There are some more architectural changes that they want to make to rustc, but they probably wouldn’t get to them in the near future even if they stopped working on `cg_clif`.

Reviewer 3: Tiny typo

```suggestion
If @bjorn3 hadn’t started working on `cg_clif` they probably wouldn’t have gained as much knowledge of the inner workings of rustc as they have now, nor contributed much to rustc. There are some more architectural changes that they want to make to rustc, but they probably wouldn’t get to them in the near future even if they stopped working on `cg_clif`.
```

Author: Fixed in e2e97f982ddc3c0ac416bb6dcc8d05eaaa35c6e1

Author: Oops, sorry for misattributing the fix in my commit message to bjorn3 instead of lqd!
